### ISSUE TITLE
X509Certificate2(Byte[]) constructor fails with certain password-less PFX-contained certificates

### ISSUE THREAD

**Issue by kristjanjogi-msft** [dotnet/runtime#87608]
### Description

The `X509Certificate2(Byte[])` constructor fails with certain password-less PFX-contained certificates. With the same certificate, the `X509Certificate2(String)` constructor (read directly from file) does not fail.

![image](https://github.com/dotnet/runtime/assets/98171324/84428600-44c4-47bc-81f9-d01104e73f93)

The PFX file does not have many iterations, like the exception's error message attempts to suggest.
![image](https://github.com/dotnet/runtime/assets/98171324/983f6742-ae1b-4a44-a935-0f19b6323b4e)

The bug affects both Windows and Linux.

The `COMPlus_Pkcs12UnspecifiedPasswordIterationLimit=-1` workaround does not work. The `new X509Certificate2(blobToImport, (string)null)` workaround does work. Workarounds taken from: https://go.microsoft.com/fwlink/?linkid=2233907

Regardless of the second workaround working, this is still a bug. It is not supposed to happen for a PFX file that has only 2000 iterations.

### Reproduction Steps

1. Produce a PFX-contained certificate that triggers this behavior. Not all certificates trigger it. The certificate I used came from Azure Key Vault. That said, not all versions of this certificate (in terms of renewals) trigger the issue. Unfortunately, I cannot share this certificate as it contains a private key.
2. Run the following code and target the certificate:
```
using System.Security.Cryptography.X509Certificates;

namespace CSharp;

internal class Program
{
	static void Main(string[] args)
	{
		var certificatePath = @"D:\certificate.pfx";

		Console.WriteLine("Attempt 1.");
		var c1 = new X509Certificate2(certificatePath);
		Console.WriteLine(c1.SubjectName.Name);

		Console.WriteLine("\nAttempt 2.");
		var c2bin = File.ReadAllBytes(certificatePath);
		var c2 = new X509Certificate2(c2bin);
		Console.WriteLine(c2.SubjectName.Name);
	}
}
```

### Expected behavior

No expection occurs.

### Actual behavior

Exception occurs.

### Configuration

```
PS D:\> dotnet --info
.NET SDK:
 Version:   7.0.304
 Commit:    7e794e2806

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.304\

Host:
  Version:      7.0.7
  Architecture: x64
  Commit:       5b20af47d9

.NET SDKs installed:
  7.0.107 [C:\Program Files\dotnet\sdk]
  7.0.304 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.18 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.18 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by stephentoub** [comment]
cc; @jeffhandley

**Comment by martincostello** [comment]
I've had a similar issue intermittently in a private app since the 7.0.304 update. I hadn't gotten around the fully investigating it yet, but here's two error messages coming from failing tests:

```

System.Security.Cryptography.CryptographicException : PKCS12 (PFX) without a supplied password has exceeded maximum allowed iterations. See https://go.microsoft.com/fwlink/?linkid=2233907 for more information.
---- System.Formats.Asn1.AsnContentException : The provided data is tagged with 'ContextSpecific' class value '4', but it should have been 'Universal' class value '16'.
```

```

System.Security.Cryptography.CryptographicException : PKCS12 (PFX) without a supplied password has exceeded maximum allowed iterations. See https://go.microsoft.com/fwlink/?linkid=2233907 for more information.
---- System.Formats.Asn1.AsnContentException : The encoded length exceeds the maximum supported by this library (Int32.MaxValue).
```

**Comment by jeffhandley** [comment]
Thank you @kristjanjogi-msft and @martincostello for reporting this.

We just updated the [KB article](https://support.microsoft.com/help/5025823) to make a few corrections to it. The `COMPlus_Pkcs12UnspecifiedPasswordIterationLimit` environment variable is not respected by .NET 6+ .NET 6+ currently only respects an `AppContext` setting to override the behavior. This was an error in our documentation. I'm sorry for the inconvenience that caused during your troubleshooting.

We suspect you are hitting **Known regression 2**, where the certificate is now misinterpreted causing its load to fail. We will fix that regression in a subsequent servicing release. When encountering that scenario, the underlying iteration count doesn't affect the behavior as .NET is misinterpreting the payload before applying the iteration count limit.

We're glad the `new X509Certificate2(blobToImport, (string)null)` workaround does work, as that is what will need to be done, at least until the known regression for misinterpreting the payload is fixed. Even once that regression is fixed though, we recommend continuing to use the `null` password workaround, as signals to the `X509Certificate2` constructor that the payload is expected to be a PFX with a `null` password, in which case the iteration count limit check will be skipped.

**Comment by krwq** [comment]
note that password overload workaround should only be used if you trust the pfx you're loading

**Comment by martincostello** [comment]
I've just read through the KB article, and also re-reviewed where my exception comes from. In my case it comes from a test where the I generate throwaway self-signed PFX file for test purposes. I've included the code for this verbatim below. I thought I'd post this here as:

- The failure mode I get is intermittent, which I guess is a function of this effectively being randomly generated.
- The certificate is inherently trusted as the code creates it and then immediately re-uses it.

It could of course be that my code to generate the certificate is slightly flawed and that's why it's now triggering this issue, but it doesn't seem to me at first sight to fit any of the scenarios in the KB that suggest how such a "bad" certificate could have been made.

```csharp
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

namespace MartinCostello.MyApp.Fixtures;

public static class CertificateFixture
{
    private static X509Certificate2? _certificate;

    public static X509Certificate2 CreateSigningCertificate()
    {
        var notBefore = DateTimeOffset.UtcNow;
        var notAfter = notBefore.AddHours(1);

        using var key = ECDsa.Create();
        var request = new CertificateRequest("CN=jwt-signing-certificate", key, HashAlgorithmName.SHA256);
        return request.CreateSelfSigned(notBefore, notAfter);
    }

    public static X509Certificate2 GetSharedSigningCertificate()
    {
        _certificate ??= CreateSigningCertificate();

        var rawData = _certificate.Export(X509ContentType.Pfx);
        return new X509Certificate2(rawData);
    }
}
```

I can easily just change the constructor in this code so it's not a big issue to apply that workaround, I just thought the scenario that created the certificate might be of interest to you.

**Comment by vcsjones** [comment]
> It could of course be that my code to generate the certificate is slightly flawed and that's why it's now triggering this issue

I don't believe that to be the case.

Given your points in 1 and 2, a durable solution here is to change the constructor call to `return new X509Certificate2(rawData, (string?)null);`.

I suspect you are also hitting the "Known regression 2" given the "random" nature of the issue you are hitting.

**Comment by martincostello** [comment]
Yep, I've just applied that change 👍

**Comment by bartonjs** [comment]
@martincostello What OS are you running that code on? (I had it in my head that whenever .NET produced a PFX, the `null` input got turned into `""`, which would avoid the circumstance... but I appear to be mistaken, at least for some OSes)

Amusingly, specifying `""` as the export password would also work around the problem for you, but since you presumably trust the PFX you're loading (from yourself), `new X509Certificate2(bytes, (string)null)` is the better workaround.

**Comment by jeffhandley** [comment]
Side note -- I'm leaving this issue open in the 8.0.0 milestone to represent that we need to fix **Known regression 2** in 8.0.0 and then backport that to .NET 6 and .NET 7.

**Comment by martincostello** [comment]
> @martincostello What OS are you running that code on?

It'll be whatever the version of Windows Server is for the `windows-latest` label in GitHub Actions (which is 2022 IIRC).

**Comment by GrabYourPitchforks** [comment]
Hi all. This issue should be resolved with today's release of .NET 7.0.8 and 6.0.19. You can download both of those from https://get.dot.net/.

Patched builds of .NET Framework 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8, and 4.8.1 are also available. See https://support.microsoft.com/kb/5028608 for more information. (We're currently busy resolving a merge conflict with the KB article contents, but the bulk of the text should have uploaded successfully. If you're receiving errors accessing the KB page, wait a few minutes and try again.)

**Comment by GrabYourPitchforks** [comment]
Fixed in 8.0 by https://github.com/dotnet/runtime/pull/87514.
